network/p2p/scoring/subscription_validator_test.go

```diff
@@ -0,0 +1,314 @@
+
+	// as peer 1 has subscribed to unknown topics, the subscription validator should return an error.
+	err := sv.CheckSubscribedToAllowedTopics(peer1)
+	require.Error(t, err)
+	require.True(t, scoring.IsInvalidSubscriptionError(err))
+}
+
+// TestSubscriptionValidator_ValidSubscription tests that when a peer has subscribed to valid
+// topics based on its Flow protocol role, the subscription validator returns no error.
+func TestSubscriptionValidator_ValidSubscriptions(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	for _, role := range flow.Roles() {
+		peer := p2pfixtures.PeerIdFixture(t)
+		// allowed channels for the role excluding the test channels.
+		allowedChannels := channels.ChannelsByRole(role).ExcludePattern(regexp.MustCompile("^(test).*"))
+		sporkID := unittest.IdentifierFixture()
+
+		allowedTopics := make([]string, 0, len(allowedChannels))
+		for _, channel := range allowedChannels {
+			allowedTopics = append(allowedTopics, channels.TopicFromChannel(channel, sporkID).String())
+		}
+
+		// peer should pass the subscription validator as it has subscribed to any subset of its allowed topics.
+		for i := range allowedTopics {
+			idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+			sp.On("GetSubscribedTopics", peer).Return(allowedTopics[:i+1])
+			require.NoError(t, sv.CheckSubscribedToAllowedTopics(peer))
+		}
+	}
+}
+
+// TestSubscriptionValidator_SubscribeToAllTopics tests that regardless of its role when a peer has subscribed to all
+// topics, the subscription validator returns an error.
+//
+// Note: this test is to ensure that the subscription validator is not bypassed by subscribing to all topics.
+// It also based on the assumption that within the list of all topics, there are guaranteed to be some that are not allowed.
+// If this assumption is not true, this test will fail. Hence, the test should be updated accordingly if the assumption
+// is no longer true.
+func TestSubscriptionValidator_SubscribeToAllTopics(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	allChannels := channels.Channels().ExcludePattern(regexp.MustCompile("^(test).*"))
+	sporkID := unittest.IdentifierFixture()
+	allTopics := make([]string, 0, len(allChannels))
+	for _, channel := range allChannels {
+		allTopics = append(allTopics, channels.TopicFromChannel(channel, sporkID).String())
+	}
+
+	for _, role := range flow.Roles() {
+		peer := p2pfixtures.PeerIdFixture(t)
+		idProvider.On("ByPeerID", peer).Return(unittest.IdentityFixture(unittest.WithRole(role)), true)
+		sp.On("GetSubscribedTopics", peer).Return(allTopics)
+		require.Error(t, sv.CheckSubscribedToAllowedTopics(peer), role)
```

Comment:
can you check the specific expected error here

Reply:
https://github.com/onflow/flow-go/pull/3291/commits/7c77d4b063f9316e53266c2ff886d4ad8fc756b6